### PR TITLE
Vectorize metric computation

### LINKED ISSUE #1345: Vectorize metric computation
### 🚀 Feature Request

Currently our metrics are calculated in per-segment fashion. We could fix it to optimize metric computation speed.

### Proposal

Add new parameter to `Metric`: `metric_fn_signature`. It could have two possible values:
- `array_to_float`: default, it expects two vectors and returns float,
- `matrix_to_array`: it expects two matrices and returns vector.

Rework functional metrics to work with new `matrix` mode by checking the input shape:
- `mape`
- `smape`
- `sign`
- `max_deviation`
- `wape`

Rework all current metrics (except for interval metrics) to work in `matrix` mode. For metrics based on `sklearn` we should use `multioutput` parameter.

To discuss:
- Could we add more metrics? For example, we could follow this logic for `Sign`, `Width`.
- Setting `multioutput` parameter manually could interfere with `self.kwargs`. What are we going to do about it?

### Test cases

- Make sure current tests pass
- Fix `test_metrics_values`
- Rework tests for functional metrics

### Additional context

Related: #1338.

### PR DESCRIPTION
**Pull request by Mr-Geekman**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
Look #1345.

## Closing issues
Closes #1345.

### CODE REVIEW

File: etna/metrics/base.py [hunk 1]
```diff
@@ -27,6 +28,23 @@ def _missing_(cls, value):
         )
 
 
+class MetricFunctionSignature(str, Enum):
+    """Enum for different metric function signatures."""
+
+    #: function should expect arrays or y_pred and y_true with length ``n_timestamps`` and return scalar
```
Comment (alex-hse-repository) [inline etna/metrics/base.py:34]: We should write about this notation in the contribution guide

Comment (alex-hse-repository) [reply etna/metrics/base.py:34]: or -> of

File: etna/metrics/base.py [hunk 2]
```diff
@@ -27,6 +28,23 @@ def _missing_(cls, value):
...
+    array_to_scalar = "array_to_scalar"
+
+    #: function should expect matrices of y_pred and y_true with shape ``(n_timestamps, n_segments)``
+    #: and return vector of length ``n_segments``
+    matrix_to_array = "matrix_to_array"
+
+    @classmethod
+    def _missing_(cls, value):
```
Comment (alex-hse-repository) [inline etna/metrics/base.py:42]: Don't we want to create base class with this method override?

File: etna/metrics/base.py [hunk 3]
```diff
@@ -66,6 +84,7 @@ def greater_is_better(self) -> Optional[bool]:
         pass
 
 
+# TODO: add tests on metric_fn_signature
```
Comment (alex-hse-repository) [inline etna/metrics/base.py:87]: ???

File: etna/metrics/base.py [hunk 4]
```diff
@@ -74,7 +93,13 @@ class Metric(AbstractMetric, BaseMixin):
     dataset and aggregates it according to mode.
     """
 
-    def __init__(self, metric_fn: Callable[..., float], mode: str = MetricAggregationMode.per_segment, **kwargs):
+    def __init__(
+        self,
+        metric_fn: Callable,
```
Comment (alex-hse-repository) [inline etna/metrics/base.py:98]: Don't we want to create an explicit signature ?
```
EtnaFuncMetric = Union[Callable[..., float], Callable[..., np.ndarray]]
```

File: etna/metrics/functional_metrics.py [hunk 5]
```diff
@@ -1,14 +1,33 @@
...
 import numpy as np
 from sklearn.metrics import mean_squared_error as mse
+from typing_extensions import assert_never
 
-ArrayLike = List[Union[float, List[float]]]
+ArrayLike = Union[float, Sequence[float], Sequence[Sequence[float]]]
 
 
-def mape(y_true: ArrayLike, y_pred: ArrayLike, eps: float = 1e-15) -> float:
+# TODO: протестировать новый режим
```
Comment (alex-hse-repository) [inline etna/metrics/functional_metrics.py:13]: ???

File: etna/metrics/functional_metrics.py [hunk 6]
```diff
@@ -1,14 +1,33 @@
...
-from typing import List
+from typing import Sequence
 from typing import Union
 
 import numpy as np
 from sklearn.metrics import mean_squared_error as mse
+from typing_extensions import assert_never
 
-ArrayLike = List[Union[float, List[float]]]
+ArrayLike = Union[float, Sequence[float], Sequence[Sequence[float]]]
```
Comment (alex-hse-repository) [inline etna/metrics/functional_metrics.py:10]: Is "float" really "ArrayLike"?

File: etna/metrics/functional_metrics.py [hunk 7]
```diff
@@ -103,20 +146,31 @@ def sign(y_true: ArrayLike, y_pred: ArrayLike) -> float:
...
+    mode:
+        Defines aggregating of multiple output values
+        (see :py:class:`~etna.metrics.functional_metrics.FunctionalMetricMode`).
+
     Returns
     -------
-    float
-        A floating point value (the best value is 0.0).
+    :
+        A non-negative floating point value (the best value is 0.0), or an array of floating point values,
```
Comment (alex-hse-repository) [inline etna/metrics/functional_metrics.py:156]: Is it non-negative here?

File: etna/metrics/functional_metrics.py [hunk 8]
```diff
@@ -42,10 +66,16 @@ def mape(y_true: ArrayLike, y_pred: ArrayLike, eps: float = 1e-15) -> float:
 
     y_true_array = y_true_array.clip(eps)
 
-    return np.mean(np.abs((y_true_array - y_pred_array) / y_true_array)) * 100
+    mode_enum = FunctionalMetricMode(mode)
```
Comment (alex-hse-repository) [inline etna/metrics/functional_metrics.py:69]: Can't we somehow avoid such code duplication? May be this enum can return correct value for an axis parameter?

File: etna/metrics/metrics.py [hunk 9]
```diff
@@ -149,7 +155,8 @@ def __init__(self, mode: str = MetricAggregationMode.per_segment, **kwargs):
         kwargs:
             metric's computation arguments
         """
-        super().__init__(mode=mode, metric_fn=mape, **kwargs)
+        mape_per_output = partial(mape, mode="per_output")
```
Comment (alex-hse-repository) [inline etna/metrics/metrics.py:158]: Don't we want to make naming consistent with sklearn? mode -> multioutput?

File: etna/metrics/metrics.py [hunk 10]
```diff
@@ -1,3 +1,5 @@
+from functools import partial
```
Comment (alex-hse-repository) [inline etna/metrics/metrics.py:1]: How are we going to deal with passing multioutput to kwargs?

File: etna/metrics/functional_metrics.py [hunk 11]
```diff
@@ -1,14 +1,33 @@
+from enum import Enum
```
Comment (alex-hse-repository) [inline etna/metrics/functional_metrics.py:1]: Can we also vectorize interval metrics?

File: tests/test_metrics/test_metrics.py [hunk 12]
```diff
@@ -210,6 +212,59 @@ def test_metrics_values(metric_class, metric_fn, train_test_dfs):
         assert value == true_metric_value
 
 
+def _create_metric_class(metric_fn, metric_fn_signature, greater_is_better):
```
Comment (alex-hse-repository) [inline tests/test_metrics/test_metrics.py:215]: Too complicated test imho

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/metrics/base.py (hunk 2 above)
Comment (Mr-Geekman) [reply etna/metrics/base.py:42]: I added suggestion about this.

File: etna/metrics/functional_metrics.py (hunk 6 above)
Comment (Mr-Geekman) [reply etna/metrics/functional_metrics.py:10]: I tried to follow smth like this: https://stackoverflow.com/a/40380014

File: tests/test_metrics/test_metrics.py (hunk 12 above)
Comment (Mr-Geekman) [reply tests/test_metrics/test_metrics.py:215]: I know, but it is the easiest way I came up with to test that old signagure and new signature give the same results. May be we could narrow this somehow.

File: etna/metrics/metrics.py (hunk 10 above)
Comment (Mr-Geekman) [reply etna/metrics/metrics.py:1]: I don't really think it should be possible to pass `multioutput` to kwargs, it doesn't make sense. Could you elaborate on your comment?

File: etna/metrics/functional_metrics.py (hunk 8 above)
Comment (Mr-Geekman) [reply etna/metrics/functional_metrics.py:69]: We could do this, but I'm not sure that it is applicable to any case. I could simplify it the other way may be and choose the `axis` parameter in this cycle.

File: etna/metrics/functional_metrics.py (hunk 11 above)
Comment (Mr-Geekman) [reply etna/metrics/functional_metrics.py:1]: There are some problems with that. First of all, interval metrics expect two columns to compute: lower quantile and higher quantile. It doesn't suit our `MetricFunctionSignature` and our current implementation of `Metric.__call__`. But there are few workarounds:
- Modify current implementation of interval metrics and not think about `Metric` and `MetricFunctionSignature`.
- Rethink `MetricFunctionSignature` and `Metric` to support more general cases of metrics.

File: etna/metrics/metrics.py (hunk 9 above)
Comment (Mr-Geekman) [reply etna/metrics/metrics.py:158]: Ok, but we should probably also discuss selected names for `mode`.

File: etna/metrics/base.py (hunk 2 above)
Comment (alex-hse-repository) [reply etna/metrics/base.py:43]: Don't see

File: etna/metrics/base.py (hunk 1 above)
Comment (alex-hse-repository) [reply etna/metrics/base.py]: What about contribution guide?

Comment (Mr-Geekman) [reply etna/metrics/base.py]: I don't think it is necessary to add this into contribution guide for now. We'll see is it necessary during reviews.

File: etna/metrics/__init__.py [hunk 13]
```diff
@@ -1,13 +1,15 @@
-from sklearn.metrics import mean_absolute_error as mae
-from sklearn.metrics import mean_squared_error as mse
-from sklearn.metrics import mean_squared_log_error as msle
-from sklearn.metrics import median_absolute_error as medae
-from sklearn.metrics import r2_score
-
 from etna.metrics.base import Metric
 from etna.metrics.base import MetricAggregationMode
+from etna.metrics.base import MetricFunctionSignature
```
Comment (alex-hse-repository) [inline etna/metrics/__init__.py:3]: Why do you thing is useful to add these classes to init?

File: etna/metrics/functional_metrics.py [hunk 14]
```diff
@@ -42,10 +69,18 @@ def mape(y_true: ArrayLike, y_pred: ArrayLike, eps: float = 1e-15) -> float:
 
     y_true_array = y_true_array.clip(eps)
 
-    return np.mean(np.abs((y_true_array - y_pred_array) / y_true_array)) * 100
+    multioutput_enum = FunctionalMetricMultioutput(multioutput)
+    if multioutput_enum is FunctionalMetricMultioutput.joint:
```
Comment (alex-hse-repository) [inline etna/metrics/functional_metrics.py:73]: I ment sth like:
```
axis = multioutput_enum.get_axis()
```

File: etna/metrics/metrics.py [hunk 15]
```diff
@@ -295,7 +306,8 @@ def __init__(self, mode: str = MetricAggregationMode.per_segment, **kwargs):
         kwargs:
             metric's computation arguments
         """
-        super().__init__(mode=mode, metric_fn=max_deviation, **kwargs)
+        max_deviation_per_output = partial(max_deviation, multioutput="per_output")
```
Comment (alex-hse-repository) [inline etna/metrics/metrics.py:309]: May be change the naming of multioutput cases to sklearn ones?  per_output -> raw_values?

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/metrics/__init__.py (hunk 13 above)
Comment (Mr-Geekman) [reply etna/metrics/__init__.py:3]: Ok, let's remove them from init. But we shouldn't forget to add them into our documentation api reference after the update.

File: etna/metrics/metrics.py (hunk 15 above)
Comment (Mr-Geekman) [reply etna/metrics/metrics.py:309]: Ok, we could do that. I don't really like the naming "raw_values", because it doesn't mean anything  for me, but it could be more intuitive for sklearn users.

Value of "joint" isn't the same as "uniform_average", are we going to do smth with it?

File: etna/metrics/functional_metrics.py (hunk 14 above)
Comment (Mr-Geekman) [reply etna/metrics/functional_metrics.py:73]: I don't find it nice to put this kind of implementation detail into public interface of enum. We could create a separate private function that does this kind of work. What do you think?

Comment (alex-hse-repository) [reply etna/metrics/functional_metrics.py:73]: Ok, let't create separate function

File: etna/metrics/metrics.py (hunk 15 above)
Comment (alex-hse-repository) [reply etna/metrics/metrics.py:309]: Why it isn't the same? mode= joint should work line sklearn default

Comment (Mr-Geekman) [reply etna/metrics/metrics.py:309]: Value "uniform average" is sklearn's default. It takes all the scores across different outputs and finds its mean value. 

Value "joint" preserves current default behaviour in etna. It works as if all matrix is flattened and one value is calculated.

So, this types produce different results for metrics that aren't calculated as mean value of some errors. For example, `max_deviation`, `wape`.

**Review by alex-hse-repository — APPROVED** [review]

**Review by alex-hse-repository — APPROVED** [review]
